src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +78,179 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
+        EventHookupSessionManager.DismissExistingSessions(cancelBackgroundTasks: false);
+
+        var task = ExecuteCommandAsync(
+            args,
+            nextHandler,
+            applicableToSpan,
+            document,
+            eventNameTask,
+            caretPoint.Value);
+        task.CompletesAsyncOperation(token);
+
+        // At this point, we've taken control, so don't send the tab into the buffer.  But do dismiss the overall
+        // session.  We no longer need it.
+        return true;
+    }
 
-        string eventHandlerMethodName = null;
-        if (EventHookupSessionManager.CurrentSession.GetEventNameTask.Status == TaskStatus.RanToCompletion)
+    private async Task ExecuteCommandAsync(
+        TabKeyCommandArgs args,
+        Action nextHandler,
+        SnapshotSpan applicableToSpan,
+        Document document,
+        Task<string> getEventNameTask,
+        SnapshotPoint initialCaretPoint)
+    {
+        try
         {
-            eventHandlerMethodName = EventHookupSessionManager.CurrentSession.GetEventNameTask.WaitAndGetResult(cancellationToken);
+            _threadingContext.ThrowIfNotOnUIThread();
+            await ExecuteCommandWorkerAsync().ConfigureAwait(false);
         }
-
-        if (eventHandlerMethodName == null ||
-            EventHookupSessionManager.CurrentSession.TextView != textView)
+        catch (OperationCanceledException)
+        {
+        }
+        catch (Exception ex) when (FatalError.ReportAndCatch(ex))
         {
-            nextHandler();
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
-            return;
         }
 
-        // This tab means we should generate the event handler method. Begin the code
-        // generation process.
-        GenerateAndAddEventHandler(textView, subjectBuffer, eventHandlerMethodName, nextHandler, cancellationToken);
-    }
-
-    private void GenerateAndAddEventHandler(ITextView textView, ITextBuffer subjectBuffer, string eventHandlerMethodName, Action nextHandler, CancellationToken cancellationToken)
-    {
-        _threadingContext.ThrowIfNotOnUIThread();
+        return;
 
-        using (Logger.LogBlock(FunctionId.EventHookup_Generate_Handler, cancellationToken))
+        async Task ExecuteCommandWorkerAsync()
         {
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
+            var textView = args.TextView;
+            var subjectBuffer = args.SubjectBuffer;
+
+            var factory = document.Project.Solution.Workspace.Services.GetRequiredService<IBackgroundWorkIndicatorFactory>();
+            using var waitContext = factory.Create(
+                textView,
+                applicableToSpan,
+                CSharpEditorResources.Generating_event);
+
+            var cancellationToken = waitContext.UserCancellationToken;
+
+            var eventHandlerMethodName = await getEventNameTask.WithCancellation(cancellationToken).ConfigureAwait(false);
```

Comment:
💭 How do we actually cancel the work from `getEventNameTask` when cancellation is requested?

Reply:
good point.  let me make that possible.